[CORL-1365] Site Specific Banning

nick-funk opened the pull request:
## What does this PR do?

Enable site specific banning for site moderators (optionally available to all other mods/users).

## What changes to the GraphQL/Database Schema does this PR introduce?

Adds siteID's to the BanHistory, BanStatus, and BanUserInput.

## How do I test this PR?

- Enable SITE_MODERATOR feature flag
- Have multiple sites under org
- Ban people as various user roles (especially a site moderator).

Note: Site mods cannot un-ban a user

tessalt committed:
allow site moderators to promote other site moderators

tessalt committed:
allow modifying roles and scopes in one mutation

nick-funk committed:
Complete the existing site moderator role management

- Site moderators can promote new site moderators
- Org mods can promote/demote commenters, staff, site mods
- Admins can do anything they want

CORL-1364

nick-funk committed:
Make it so user's can't reassign their own roles

CORL-1364

nick-funk committed:
Preliminarily hook up back-end for site-specific bans

CORL-1365

nick-funk committed:
Update the front end to show site-specific ban features

CORL-1365

nick-funk committed:
Separate active vs site specific ban checks

CORL-1365

nick-funk committed:
Preliminarily handle removing a ban from a user

CORL-1365

nick-funk committed:
Refactor site bans into separate flows

Also refactor consolidation to compute banned site ids
from the history.

CORL-1365

nick-funk committed:
Create removeUserSiteBan to handle site un-ban flow

CORL-1365

nick-funk committed:
Enforce site selection ban for site moderators

Optional choice for other moderators and admins

CORL-1365

wyattjoh commented on src/core/client/admin/components/UserRole/PromoteUserRoleMutation.ts:
```diff
@@ -0,0 +1,45 @@
+        },
+      },
+    })
+);
```
This looks like it was copied over from the other PR!

nick-funk replied:
I branched off the other PR, so yes, you're very correct haha

wyattjoh commented on src/core/server/graph/resolvers/UserStatus.ts:
```diff
@@ -53,10 +81,14 @@ export const UserStatus: Required<GQLUserStatusTypeResolver<
+    const banStatus = user.consolidateUserBanStatus(ban, ctx.now, ctx.site?.id);
+    return {
+      ...banStatus,
+      sites: await loadSites(ctx, banStatus.siteIDs),
```
This could instead be written as:

```suggestion
      sites: await ctx.loaders.Sites.site.loadMany(banStatus.siteIDs),
```

Because the `Site` resolver will automatically handle adding the `canModerate` and `topStories` for you.

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -2042,6 +2047,11 @@ type BanStatus {
+  """
+  sites is a list of sites that the user has been site banned on.
+  """
+  sites: [Site!]!
```
This should instead be:

```suggestion
  sites: [Site!]
```

As the user may not be banned on any sites.

wyattjoh reviewed (commented):
Looks like some copy/paste from the other PR!

nick-funk committed:
Use loadSites instead of ctx deep loader

CORL-1365

Co-authored-by: Wyatt Johnson <wyattjoh@gmail.com>

nick-funk committed:
Make sites array required instead of optional

CORL-1365

Co-authored-by: Wyatt Johnson <wyattjoh@gmail.com>

nick-funk committed:
User GQLUSER_ROLE_RL generated types for roles

CORL-1364

nick-funk committed:
Simplify viewer dependencies in UserRoleChange

CORL-1364

nick-funk committed:
Clean up BanStatus resolvers and directive compute

CORL-1365

nick-funk committed:
Fix ban state checks to account for sites

CORL-1365

nick-funk committed:
Disable removing a ban as a site mod

CORL-1365

nick-funk committed:
Allow site mods to call ban mutation

CORL-1365

nick-funk committed:
Perform site-specific ban when moderating in stream

CORL-1365

nick-funk committed:
Check siteIDs length in canModerate resolver

When evaluating canModerate with siteIDs on moderators
account for the case where an org mod has an empty siteIDs
moderation scope. If their siteIDs list is empty, then
the moderator is not scoped and can moderate.

CORL-1365

nick-funk committed:
Ensure site mods provide a site ID when banning

CORL-1365

nick-funk committed:
Improve site specific ban form validation

CORL-1365

nick-funk committed:
Remove unused PromoteUserRoleMutation

CORL-1365

nick-funk commented:
Destroyed by failed rebase and bad prior merging/changes. Closing.
